src/VisualStudio/CSharp/Test/ProjectSystemShim/CPS/SourceFileHandlingTests.cs

```diff
@@ -57,5 +58,350 @@ public void AddRemoveAdditionalFile_CPS()
+                project.AddSourceFile(sourceFileFullPath2);
+                project.AddSourceFile(sourceFileFullPath3);
+                project.AddSourceFile(sourceFileFullPath4);
+                project.AddSourceFile(sourceFileFullPath5);
+
+                project.RemoveSourceFile(sourceFileFullPath2);
+                project.AddSourceFile(sourceFileFullPath2);
+
+                project.RemoveSourceFile(sourceFileFullPath4);
+                project.AddSourceFile(sourceFileFullPath4);
+
+                project.ReorderSourceFiles(new[] { sourceFileFullPath1, sourceFileFullPath2, sourceFileFullPath3, sourceFileFullPath4, sourceFileFullPath5 });
+
+                var documents = GetCurrentDocuments().ToArray();
+
+                Assert.Equal(documents[0].FilePath, sourceFileFullPath1, StringComparer.OrdinalIgnoreCase);
+                Assert.Equal(documents[1].FilePath, sourceFileFullPath2, StringComparer.OrdinalIgnoreCase);
+                Assert.Equal(documents[2].FilePath, sourceFileFullPath3, StringComparer.OrdinalIgnoreCase);
+                Assert.Equal(documents[3].FilePath, sourceFileFullPath4, StringComparer.OrdinalIgnoreCase);
+                Assert.Equal(documents[4].FilePath, sourceFileFullPath5, StringComparer.OrdinalIgnoreCase);
+            }
+        }
+
+        [WpfFact]
+        [Trait(Traits.Feature, Traits.Features.ProjectSystemShims)]
+        public void ReorderSourceFilesBatch_CPS()
+        {
+            using (var environment = new TestEnvironment())
+            using (var project = CreateCSharpCPSProject(environment, "project1"))
+            {
+                IEnumerable<Document> GetCurrentDocuments() => environment.Workspace.CurrentSolution.Projects.Single().Documents;
+
+                Assert.Empty(GetCurrentDocuments());
+
+                var sourceFileFullPath1 = @"c:\source1.cs";
+                var sourceFileFullPath2 = @"c:\source2.cs";
+                var sourceFileFullPath3 = @"c:\source3.cs";
+                var sourceFileFullPath4 = @"c:\source4.cs";
+                var sourceFileFullPath5 = @"c:\source5.cs";
+
+                // Add a file outside the batch.
+                project.AddSourceFile(sourceFileFullPath2);
+
+                project.StartBatch();
+                project.AddSourceFile(sourceFileFullPath1);
+                project.AddSourceFile(sourceFileFullPath3);
+                project.AddSourceFile(sourceFileFullPath4);
+                project.AddSourceFile(sourceFileFullPath5);
+
+                // Removing path2 to test removal of a file the actual internal project state has changed outside of the batch.
+                project.RemoveSourceFile(sourceFileFullPath2);
+
+                // Removing path4 to test remove of a file when it was also added in a batch.
+                project.RemoveSourceFile(sourceFileFullPath4);
+
+                project.ReorderSourceFiles(new[] { sourceFileFullPath1, sourceFileFullPath3, sourceFileFullPath5 });
```

Comment:
Note: this is the order the files already would have been in, so this test isn't really testing that the reorder did anything. You might have this covered in the next test though; up to you.

Reply:
I'll do it here as well just so we cover everything.